### ISSUE TITLE
twisted.internet.endpoints.clientFromString does not honor IPv6 addresses (always returns TCP4ClientEndpoint)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8034]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| posita reported|
|-|-|
|Trac ID|trac#8034|
|Type|defect|
|Created|2015-09-28 19:08:41Z|

This relates to (or extends) [#5695](https://github.com/twisted/twisted/issues/5695). Consider:

```
>>> from twisted.names import client
>>> d = client.getHostByName('www.example.com')
>>> d
<Deferred at 0x10b4f6c68 current result: '2606:2800:220:1:248:1893:25c8:1946'>
```

But, a `TCP4ClientEndpoint` is still created by `clientFromString`:

```
>>> from twisted.internet.endpoints import clientFromString
>>> clientFromString(reactor, "tcp:host=www.example.com:80")
<twisted.internet.endpoints.TCP4ClientEndpoint object at 0x10b46e510>
```

IPv6 bracket notation does not appear to be supported:

```
>>> clientFromString(reactor, "tcp:host=[2606:2800:220:1:248:1893:25c8:1946]:100")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/.../twisted/internet/endpoints.py", line 1749, in clientFromString
    return _endpointClientFactories[name](reactor, **kwargs)
TypeError: __init__() takes at least 4 arguments (3 given)
>>> clientFromString(reactor, "tcp:host=[2606:2800:220:1:248:1893:25c8:1946]:port=100")
<twisted.internet.endpoints.TCP4ClientEndpoint object at 0x10b4d8fd0>
>>> _._host
'[2001'
```

The (non-standard) escape notation helps designate the host, but the endpoint is still the wrong type:

```
>>> clientFromString(reactor, "tcp:host=2606\:2800\:220\:1\:248\:1893\:25c8\:1946:100")
<twisted.internet.endpoints.TCP4ClientEndpoint object at 0x10ae90790>
>>> _._host
'2606:2800:220:1:248:1893:25c8:1946'
>>> ep._port
100
>>> clientFromString(reactor, "tcp:host=2606\:2800\:220\:1\:248\:1893\:25c8\:1946:port=100")
<twisted.internet.endpoints.TCP4ClientEndpoint object at 0x10b4a1c10>
>>> _._host
'2606:2800:220:1:248:1893:25c8:1946'
>>> ep._port
100
```

<details><summary>Searchable metadata</summary>

```
trac-id__8034 8034
type__defect defect
reporter__posita posita
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__core core
keywords__None None
time__1443467321068370 1443467321068370
changetime__1445922370947471 1445922370947471
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

TCP4ClientEndpoint will actually work just fine with IPv6 literals.  (However, `tcp:` should really return a `HostnameEndpoint`.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

We use keywords for workflow, not random tag associations, so deleting "endpoint" and "ipv6"

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set status to `closed`|
|-|-|

Here's an example of what I mean:

```
>>> from twisted.internet.endpoints import clientFromString
>>> endpoint = clientFromString(reactor, r"tcp:host=\:\:1:port=22")
>>> from twisted.internet.protocol import Protocol
>>> class Appender(Protocol):
...     def dataReceived(self, data):
...         l.append(data)
... 
>>> l = []
>>> from twisted.internet.protocol import Factory
>>> endpoint.connect(Factory.forProtocol(Appender))
<Deferred at 0x112718170>
Waiting on Deferred #0...
Deferred #0 Fired:
<__main__.Appender instance at 0x10ecc0830>
>>> l
['SSH-2.0-OpenSSH_6.9\r\n']
>>> r = _.result
>>> r.transport.getPeer()
IPv6Address(TCP, '::1', 22)
```

`IPv4Address` in fact just calls `connectTCP`, by design, so that IPv6 connections were always possible; `connectTCP` checks for IPv6 address literals. `IPv6Address` was introduced slightly later, all _it_ does is hook up _name resolution_ to IPv6-only hosts.

In the case of an IPv6 literal, client endpoints actually work as designed, except for slightly awkward naming and inconvenient escaping syntax.  However, there is still no way to construct a string endpoint which constructs a `HostnameEndpoint`; `tcp:` should be changed to do that.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|posita commented|
|-|-|

> TCP4ClientEndpoint will actually work just fine with IPv6 literals. (However, `tcp:` should really return a `HostnameEndpoint`.)

Note to self: [#7387](https://github.com/twisted/twisted/issues/7387) covers that.
